sound/soc/sof/ipc4-topology.c

```diff
@@ -1589,8 +1590,48 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		return -EINVAL;
 	}
 
+	/* check if all input formats are the same */
+	single_input_format = sof_ipc4_copier_is_single_format(sdev,
+								available_fmt->input_pin_fmts,
+								available_fmt->num_input_formats);
+
+	/*
+	 * in the case of capture, if all the input formats are the same, update the reference
+	 * params to use the only available format when searching for a match. This is needed
+	 * for cases when the DAI copier has multiple input audio formats for variable channel
+	 * counts and rates but the sample format is the same across all of them.
+	 */
+	if (single_input_format && dir == SNDRV_PCM_STREAM_CAPTURE) {
+		struct sof_ipc4_audio_format *in_fmt;
+		snd_pcm_format_t snd_fmt;
+		struct snd_mask *m;
+		int in_ref_valid_bits;
+
+		in_fmt = &available_fmt->input_pin_fmts[0].audio_fmt;
+		in_ref_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(in_fmt->fmt_cfg);
+
+		switch (in_ref_valid_bits) {
+		case 16:
+			snd_fmt = SNDRV_PCM_FORMAT_S16_LE;
+			break;
+		case 24:
+			snd_fmt = SNDRV_PCM_FORMAT_S24_LE;
+			break;
+		case 32:
+			snd_fmt = SNDRV_PCM_FORMAT_S32_LE;
+			break;
+		default:
+			dev_err(sdev->dev, "invalid PCM valid_bits %d\n", in_ref_valid_bits);
+			return -EINVAL;
+		}
+
+		m = hw_param_mask(&ref_params, SNDRV_PCM_HW_PARAM_FORMAT);
+		snd_mask_none(m);
+		snd_mask_set_format(m, snd_fmt);
+	}
+
 	/* set input and output audio formats */
-	ret = sof_ipc4_init_input_audio_fmt(sdev, swidget, &copier_data->base_config, ref_params,
+	ret = sof_ipc4_init_input_audio_fmt(sdev, swidget, &copier_data->base_config, &ref_params,
 					    available_fmt);
```

Comment:
I don't get how this works...
When we have single format, this all looks like a NOP: the sof_ipc4_init_input_audio_fmt() ignores the ref_params in that case, so the whole `if (single_input_format && dir == SNDRV_PCM_STREAM_CAPTURE)` is just done for nothing.
Line 1199

I'm sure I'm missing something...

We try to narrow the  format support if only one format is supported (which should by itself should narrow it?), but what is expect if we have two formats supported, like S24_LE and S32_LE? We are no masking out the S16_LE?